scalafix-core/shared/src/main/scala/scalafix/internal/rule/DisableSyntax.scala

```diff
@@ -35,38 +37,145 @@ final case class DisableSyntax(
-        case token @ Token.Xml.Start() if config.noXml =>
-          error("noXml", token)
-      }.toSeq
-    val treeLintMessages =
-      ctx.tree.collect {
-        case t @ mod"+" if config.noCovariantTypes =>
+    regexLintMessages.result()
+  }
+
+  private def checkTokens(ctx: RuleCtx): Seq[LintMessage] = {
+    ctx.tree.tokens.collect {
+      case token @ Keyword(keyword) if config.isDisabled(keyword) =>
+        errorCategory
+          .copy(id = s"keywords.$keyword")
+          .at(s"$keyword is disabled", token.pos)
+      case token @ Token.Semicolon() if config.noSemicolons =>
+        error("noSemicolons", token)
+      case token @ Token.Tab() if config.noTabs =>
+        error("noTabs", token)
+      case token @ Token.Xml.Start() if config.noXml =>
+        error("noXml", token)
+    }
+  }
+
+  private def checkTree(ctx: RuleCtx): Seq[LintMessage] = {
+    object AbstractWithVals {
+      def unapply(t: Tree): Option[List[Defn.Val]] = {
+        val stats = t match {
+          case Defn.Class(mods, _, _, _, templ)
+              if mods.exists(_.is[Mod.Abstract]) =>
+            templ.stats
+          case Defn.Trait(_, _, _, _, templ) => templ.stats
+          case _ => List.empty
+        }
+        val vals = stats.flatMap {
+          case v: Defn.Val => Some(v)
+          case _ => None
+        }
+        if (vals.isEmpty) None else Some(vals)
+      }
+    }
+
+    object WithMethods {
+      def unapply(t: Tree): Option[List[Defn.Def]] = {
+        val stats = t match {
+          case Defn.Class(_, _, _, _, templ) => templ.stats
+          case Defn.Trait(_, _, _, _, templ) => templ.stats
+          case Term.NewAnonymous(templ) => templ.stats
+          case _ => List.empty
+        }
+        val methods = stats.flatMap {
+          case d: Defn.Def => Some(d)
+          case _ => None
+        }
+        if (methods.isEmpty) None else Some(methods)
+      }
+    }
+
+    def hasDefaultArgs(d: Defn.Def): Boolean =
+      d.paramss.exists(_.exists(_.default.isDefined))
+
+    def hasNonImplicitParam(d: Defn.Def): Boolean =
+      d.paramss.exists(_.exists(_.mods.forall(!_.is[Mod.Implicit])))
+
+    ctx.tree.collect {
+      case t @ mod"+" if config.noCovariantTypes =>
+        Seq(
           errorCategory
             .copy(id = "covariant")
             .at(
               "Covariant types could lead to error-prone situations.",
               t.pos
             )
-        case t @ mod"-" if config.noContravariantTypes =>
+        )
+      case t @ mod"-" if config.noContravariantTypes =>
+        Seq(
           errorCategory
             .copy(id = "contravariant")
             .at(
               "Contravariant types could lead to error-prone situations.",
               t.pos
             )
-      }
+        )
+      case t @ WithMethods(methods)
+          if methods.exists(hasDefaultArgs) && config.noDefaultArgs =>
+        methods
+          .filter(hasDefaultArgs)
+          .map(
```

Comment:
Pro tip, scalafmt dislikes large callbacks with parentheses so it looks a lot nicer with curly braces

Reply:
some people don't like curly braces for single expressions but I personally disagree!